DirectProgramming/DPC++FPGA/include/pipe_utils.hpp

```diff
@@ -0,0 +1,264 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#ifndef __PIPE_ARRAY_HPP__
+#define __PIPE_ARRAY_HPP__
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+#include <utility>
+
+/*
+
+This header defines the following utilities for use with pipes in DPC++ FPGA
+designs.
+
+1. PipeArray
+
+      Create a collection of pipes that can be indexed like an array.
+
+      template <class Id,          // identifier for the pipe array
+                typename BaseTy,   // type to write/read for each pipe
+                size_t min_depth,  // minimum capacity of each pipe
+                size_t... dims     // depth of each dimension in the array
+                                   // any number of dimensions are supported
+                >
+      struct PipeArray
+
+      Example usage:
+    
+      class PipeArrayId;
+      constexpr int min_depth = 0;
+      constexpr int num_pipes = 4;
+      using MyPipeArray = PipeArray<PipeArrayId, int, min_depth, num_pipes>;
+      ...
+      constexpr int pipe_idx = 1;
+      MyPipeArray::PipeAt<pipe_idx>::read(); 
+
+2. PipeDuplicator
+
+      Fan-out a single pipe write to multiple pipe instances,
+      each of which will receive the same data.
+      A blocking write will perform a blocking write to each pipe.
+      A non-blocking write will perform a non-blocking write to each pipe,
+      and set success to true only if ALL writes were successful.
```

Comment:
I think it's worth mentioning the 'null pipe' case here, as it can be very useful as a stub for an unwanted pipe.  How about:

Note that the special case of 0 pipe instances is supported, which can be useful as a stub for writes to pipes that are not needed in your particular design.

Reply:
I was thinking that we could update this file with a wrapper class to cover that use case cleanly after this initial change.

Reply:
Why not include the comment for now until you make that wrapper class?

Reply:
Done